docker/Dockerfile.build

```
@@ -1,56 +1,88 @@
-# Builder image
-FROM rust:buster as builder
+# This steps implement cargo-chef for docker layer caching
+# We are using four stages:
+# - chef: installs cargo-chef
+# - planner: computes the recipe file
+# - builder: caches our dependencies and builds the binary
+# - tester: builds and run tests
+# - runtime: is our runtime environment
+FROM rust:bullseye as chef
+RUN cargo install cargo-chef --locked
+WORKDIR /app
+
+FROM chef AS planner
+COPY . .
+RUN cargo chef prepare --recipe-path recipe.json
+
+FROM chef AS builder
+COPY --from=planner /app/recipe.json recipe.json
 
 RUN apt-get update && \
-	apt-get install -y --no-install-recommends \
-	make cmake g++ gcc llvm libclang-dev clang ca-certificates
+    apt-get install -y --no-install-recommends \
+    llvm libclang-dev clang ca-certificates && \
+    rm -rf /var/lib/apt/lists/* /tmp/*
 
-RUN mkdir /zebra
-WORKDIR /zebra
+# Optimize builds. In particular, regenerate-stateful-test-disks.yml was reaching the
+# GitHub Actions time limit (6 hours), so we needed to make it faster.
+ENV RUSTFLAGS -O
+ENV CARGO_HOME /app/.cargo/
+# Build dependencies - this is the caching Docker layer!
+RUN cargo chef cook --release --features enable-sentry --recipe-path recipe.json
 
-ARG SHORT_SHA
-ENV SHORT_SHA $SHORT_SHA
+# Build zebra
+ARG RUST_BACKTRACE=1
+ENV RUST_BACKTRACE ${RUST_BACKTRACE}
 
-ENV RUST_BACKTRACE full
-ENV CARGO_HOME /zebra/.cargo/
+COPY . .
+RUN cargo build --locked --jobs 2 --release --features enable-sentry --bin zebrad
```

The default for `jobs` is the number of detected processor cores.
If we set a fixed default, the build won't run faster on larger machines.

Is the detection failing here?

```suggestion
RUN cargo build --locked --release --features enable-sentry --bin zebrad
```

---

Also `--locked` is the default for `cargo build`, so it's not needed here.
(But unfortunately it's needed for `cargo install`.)

---

When I was researching cargo build, I stumbled with these two issues:
https://github.com/rust-lang/cargo/issues/4566
https://github.com/rust-lang/cargo/issues/8207

This made think that adding --locked had a different behavior

---

You're right, `--locked` works slightly differently to what I said.

For `cargo build`:
> After the first time it runs, the result is stored in the Cargo.lock file. Subsequent commands will run the resolver, keeping dependencies locked to the versions in Cargo.lock *if it can*.

https://doc.rust-lang.org/cargo/reference/resolver.html#dependency-updates

But for `cargo install`:
> By default, the Cargo.lock file that is included with the package will be ignored.

https://doc.rust-lang.org/cargo/commands/cargo-install.html#description

So if we want to make sure we're building with the exact dependencies, we should use `--locked` everywhere.